In [1]:
# working dir
import os
import sys

cwd = os.getcwd()
root_dir = os.path.dirname(os.path.dirname(cwd))
sys.path.append(root_dir)

print(root_dir)

c:\Users\david\Desktop\Main\03_Uni\WWI21DSA\02_Vorlesungen\06_Projektrealisierung\Projektrealisierung


In [2]:
# Imports and settings
import pandas as pd
import random

random.seed(42)

pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)
pd.set_option("display.max_colwidth", None)

In [14]:
# Load data
abcd = pd.read_csv("../../src/data/ABCD_tripfiles_preprocessed.csv")
mnop = pd.read_csv("../../src/data/MNOP_tripfiles_preprocessed.csv")
zyxw = pd.read_csv("../../src/data/ZYXW_tripfiles_preprocessed.csv")

C:\Users\david\AppData\Local\Temp\ipykernel_27460\136873643.py:3: DtypeWarning: Columns (12,13,14,15,16,17,18,19,20,21,22,23,25,26) have mixed types. Specify dtype option on import or set low_memory=False.
  mnop = pd.read_csv("../../src/data/MNOP_tripfiles_preprocessed.csv")
C:\Users\david\AppData\Local\Temp\ipykernel_27460\136873643.py:4: DtypeWarning: Columns (12,13,14,15,16,17,18,19,20,21,22,24,25,26,27) have mixed types. Specify dtype option on import or set low_memory=False.
  zyxw = pd.read_csv("../../src/data/ZYXW_tripfiles_preprocessed.csv")


In [4]:
# Subset of data
random_flights = random.sample(abcd["header_id"].tolist(), 250)
abcd_subset = abcd[abcd["header_id"].isin(random_flights)]

In [5]:
meta = ['id', 'creation_time', 'airline_code', 'flight_number', 'flight_date',
       'departure_airport', 'user_name', 'action_name']

message = ['header_line', 'entry_details']

header_cols = ['header_category', 'header_id']

extracted_cols = ['scheduleState',
       'departureAirport', 'departureTime', 'arrivalAirport', 'arrivalTime',
       'aircraftRegistration', 'aircraftSubtype', 'aircraftVersion',
       'serviceType', 'nonOperational', 'flifoCancel', 'sequenceNumber',
       'bestTimes', 'mvtActual']

weight_cols = ['START_WI_weight', 'START_WI_index',
       'DO_WI_weight', 'DO_WI_index', 'PAX_WI_weight', 'PAX_WI_index',
       'TOTAL_DEADLOAD_WI_weight', 'TOTAL_DEADLOAD_WI_index', 'TOTAL_LOAD_WI',
       'TOTAL_TRAFFIC_LOAD', 'FUEL_INDEX', 'AZFW', 'ATOW', 'ALAW', 'ATXW',
       'AFT_LIMIT_ZFW', 'FWD_LIMIT_ZFW', 'AFT_LIMIT_TOW', 'FWD_LIMIT_TOW',
       'AFT_LIMIT_LAW', 'FWD_LIMIT_LAW', 'LIZFW', 'LITOW', 'LILAW',
       'MAC_AT_ZFW', 'MAC_AT_TOW', 'MAC_AT_LAW', 'DEADLOAD_MAC', 'UNDERLOAD',
       'LIMITING_WEIGHT', 'ALLOWED_TOW', 'ALLOWED_ZFW', 'ALLOWED_LAW',
       'ALLOWED_TXW', 'STABTO', 'OPTIMAL_TRIM', 'IDEAL_ADDITIONAL_LOAD_AFT',
       'IDEAL_ADDITIONAL_LOAD_FWD', 'TAIL_TIPPING_WI_weight',
       'TAIL_TIPPING_WI_index', 'TAIL_TIPPING_INDEX_EXCEEDED',
       'FWD_MOVABLE_PAX', 'AFT_MOVABLE_PAX', 'INDEX_OUT_OF_BALANCE',
       'LOAD_TO_AFT', 'LOAD_TO_FWD', 'ESTIMATED_TRAFFIC_LOAD', 'ESTIMATED_ZFW',
       'DELTA_ZFW', 'ZFW_TOLERANCE_EXCEEDED']


In [6]:
# Zeigt wie oft die einzelnen actions pro flug vorkommen
action_occurences = abcd_subset.drop_duplicates(subset=['header_id', 'action_name'])
action_occurences['action_name'].value_counts()

action_name
CalculateWeightAndTrimAction           234
StorePaxDataAction                     175
CheckinMsgProcessor                    147
UpdateCrewDataAction                    30
ASMMsgProcessor                         27
UpdateLoadTableAction                   25
UpdateFlightAction                      20
AssignLCCAction                         20
StoreRegistrationAndConfigurationAc     18
ResetLoadingListRecordsAction           15
CrewMsgProcessor                        12
EstimateStorePaxDataAction              11
PAXBOOKINGINMsgProcessor                 9
TransferCheckinDataAction                8
ChangeFlightLegStateAction               6
CreatePostDepartureMessagesAction        6
UpdateHistoricalFlightLegInfoAction      5
CloseLegAction                           5
InternalCreateLoadingInstructionAct      5
SendPostDepartureMessagesAction          5
InternalCreateLoadsheetAction            5
CreateLoadingInstructionAction           4
UpdateSupplementaryInfoAction            4

### >> Finding 1: Wenn man header_id als index nimmt -> Von den 250 flights im subset hatten 234 einen Eintrag für die Action CalculateWeightAndTrimAction. Die anderen Aktionen kamen seltener vor und stehen dementsprechend für weniger flüge zur Verfügung.

In [7]:
# create makeshift flight identifier
abcd["temp"] = abcd["departureAirport"] + abcd["arrivalAirport"] + abcd["departureTime"] + abcd["arrivalTime"]

In [8]:
# Wie viele Flüge gäbe es, bei den unterschiedlichen identifizierungsansätzen
print(
    len(abcd),

    abcd["id"].nunique(),
    abcd["header_id"].nunique(),
    abcd["flight_number"].nunique(),
    abcd["temp"].nunique()
)

709613 709613 121282 163 2379


In [9]:
# Stichprobenartig einen "flug" (nach temp index) anschauen
temp_table = abcd[abcd["temp"]=="BLRDEL2024-05-01T23:00:00.000Z2024-05-02T01:50:00.000Z"]

temp_table[["header_id", "header_category"] + meta + ['departureAirport', 'departureTime', 'arrivalAirport', 'arrivalTime',
       'aircraftRegistration', 'aircraftSubtype', 'aircraftVersion']].sort_values("creation_time")

header_id header_category        id        creation_time  \
207660  9d1a19e10783012d           saved  33846020  2024-04-30 22:14:13   
207661  9d1a19e10783012d        received  33845797  2024-04-30 22:14:13   
207662  9d1a19e10783012d        received  33845795  2024-04-30 22:14:13   
480043  9d1a19e10783012d        received  33845792  2024-04-30 22:14:13   
37506   9d1a19e10783012d        received  33845798  2024-04-30 22:14:13   
37505   9d1a19e10783012d        received  33845790  2024-04-30 22:14:13   
480044  9d1a19e10783012d        received  33845791  2024-04-30 22:14:13   
25579   9d1a19e10783012d           saved  33846025  2024-04-30 22:14:13   
480045  9d1a19e10783012d           saved  33846024  2024-04-30 22:14:13   
25577   9d1a19e10783012d           saved  33845799  2024-04-30 22:14:13   
507245  9d1a19e10783012d        received  33845796  2024-04-30 22:14:13   
507246  9d1a19e10783012d           saved  33846026  2024-04-30 22:14:13   
507244  9d1a19e10783012d        received  33845789  2024-04-30 22:14:13   
13334   9d1a19e10783012d           saved  33846027  2024-04-30 22:14:13   
530022  9d1a19e10783012d        received  33845794  2024-04-30 22:14:13   
530023  9d1a19e10783012d           saved  33846022  2024-04-30 22:14:13   
558914  9d1a19e10783012d        received  33845793  2024-04-30 22:14:13   
558915  9d1a19e10783012d           saved  33846023  2024-04-30 22:14:13   
252     9d1a19e10783012d        received  33845788  2024-04-30 22:14:13   
207659  9d1a19e10783012d           saved  33846021  2024-04-30 22:14:13   
509234   3f326d134d9de45        received  33902389  2024-05-01 09:14:42   
569788   3f326d134d9de45        received  33902386  2024-05-01 09:14:42   
482229   3f326d134d9de45           saved  33902397  2024-05-01 09:14:42   
482228   3f326d134d9de45        received  33902385  2024-05-01 09:14:42   
482227   3f326d134d9de45        received  33902384  2024-05-01 09:14:42   
509235   3f326d134d9de45           saved  33902399  2024-05-01 09:14:42   
531966   3f326d134d9de45        received  33902387  2024-05-01 09:14:42   
531967   3f326d134d9de45           saved  33902395  2024-05-01 09:14:42   
509233   3f326d134d9de45        received  33902382  2024-05-01 09:14:42   
0        3f326d134d9de45        received  33902381  2024-05-01 09:14:42   
203481   3f326d134d9de45        received  33902388  2024-05-01 09:14:42   
14226    3f326d134d9de45           saved  33902860  2024-05-01 09:14:42   
203482   3f326d134d9de45        received  33902390  2024-05-01 09:14:42   
25580    3f326d134d9de45           saved  33902392  2024-05-01 09:14:42   
38320    3f326d134d9de45        received  33902383  2024-05-01 09:14:42   
25578    3f326d134d9de45           saved  33902398  2024-05-01 09:14:42   
569789   3f326d134d9de45           saved  33902396  2024-05-01 09:14:42   
203480   3f326d134d9de45           saved  33902393  2024-05-01 09:14:42   
203479   3f326d134d9de45           saved  33902394  2024-05-01 09:14:42   
38321    3f326d134d9de45        received  33902391  2024-05-01 09:14:42   
4679    59aa480eb5f2c287        received  33928599  2024-05-01 15:01:31   
14513   59aa480eb5f2c287           saved  33930342  2024-05-01 15:01:31   
281162  59aa480eb5f2c287           saved  33930341  2024-05-01 15:01:31   
281161  59aa480eb5f2c287        received  33930340  2024-05-01 15:01:31   
8188    795a5ebf29415013        received  33938820  2024-05-01 18:46:17   
14628   795a5ebf29415013           saved  33938823  2024-05-01 18:46:17   
281327  795a5ebf29415013        received  33938821  2024-05-01 18:46:17   
281328  795a5ebf29415013           saved  33938822  2024-05-01 18:46:17   
8189    e7e7b864573365f6        received  33941480  2024-05-01 21:00:26   
14726   e7e7b864573365f6           saved  33941483  2024-05-01 21:00:26   
281418  e7e7b864573365f6           saved  33941482  2024-05-01 21:00:26   
281417  e7e7b864573365f6        received  33941481  2024-05-01 21:00:26   
14751    628a38f44e665b3           saved

In [13]:
# Zeigt wie oft die einzelnen actions pro flug vorkommen
abcd.drop_duplicates(subset=['flight_id', 'action_name'])['action_name'].value_counts()

action_name
ASMMsgProcessor                        2379
UpdateFlightAction                     1997
AssignLCCAction                        1997
StoreRegistrationAndConfigurationAc    1996
UpdateCrewDataAction                   1992
CalculateWeightAndTrimAction           1992
ResetLoadingListRecordsAction          1489
UpdateLoadTableAction                  1489
TransferCargoAction                     964
ChangeFlightLegStateAction              944
StorePaxDataAction                      741
FuelDataInitializer                      71
CloseLegAction                            1
SendLoadsheetAction                       1
StorePaxDataGuiAction                     1
StoreCKIAverageWeightAction               1
CheckinMsgProcessor                       1
CargoFinalAction                          1
AssignUnassignViewAction                  1
SendLoadingInstructionAction              1
UpdateHistoricalFlightLegInfoAction       1
StoreAircraftDataAction                   1
UpdateFuelDataAction

### >> Finding 2: Wenn man abcd["departureAirport"] + abcd["arrivalAirport"] + abcd["departureTime"] + abcd["arrivalTime"] als index nimmt -> So ergeben sich für den ersten Datensatz (ABCD) 2379 Flüge. 6 Aktionen kommen in >1900 Fällen vor. 